packages/next/src/lib/metadata/resolve-metadata.tsx

```diff
@@ -0,0 +1,149 @@
+import type {
+  Metadata,
+  ResolvedMetadata,
+  ResolvingMetadata,
+} from './types/metadata-interface'
+import type { AbsoluteTemplateString } from './types/metadata-types'
+
+import React from 'react'
+import fs from 'fs-extra'
+import path from 'path'
+import { DEFAULT_METADATA } from './constant'
+import { resolveOpenGraph } from './resolve-opengraph'
+import { resolveTitle } from './resolve-title'
+import { elementsFromResolvedOpenGraph } from './generate/opengraph'
+
+type Item = {
+  type: 'layout' | 'page'
+  mod: () => Promise<{
+    metadata?: Metadata
+    generateMetadata?: (
+      props: any,
+      parent: ResolvingMetadata
+    ) => Promise<Metadata>
+  }>
+}
+
+// Merge the source metadata into the resolved target metadata.
+function merge(source: Metadata, target: ResolvedMetadata) {
+  const stashedTitle: AbsoluteTemplateString = target.title || {
+    absolute: '',
+    template: '%s',
+  }
+  const stashedOpenGraphTitle: AbsoluteTemplateString = target.openGraph
+    ?.title || {
+    absolute: '',
+    template: '%s',
+  }
+  const stashedTwitterTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: '%s',
+  }
+
+  for (const key_ in source) {
+    const key = key_ as keyof Metadata
+
+    if (key === 'other') {
+      target.other = { ...target.other, ...source.other }
+    } else if (key === 'title') {
+      resolveTitle(stashedTitle, source.title)
+      target.title = stashedTitle
+    } else if (key === 'openGraph') {
+      if (source.openGraph && 'title' in source.openGraph) {
+        resolveTitle(stashedOpenGraphTitle, source.openGraph.title)
+      }
+      if (typeof source.openGraph !== 'undefined') {
+        target.openGraph = {
+          ...resolveOpenGraph(source.openGraph),
+          title: stashedOpenGraphTitle,
+        }
+      } else {
+        target.openGraph = null
+      }
+    } else if (key === 'twitter') {
+      if (source.twitter && 'title' in source.twitter) {
+        resolveTitle(stashedTwitterTitle, source.twitter.title)
+      }
+      if (typeof source.twitter !== 'undefined') {
+        target.twitter = { ...source.twitter, title: stashedTwitterTitle }
+      } else {
+        target.twitter = null
+      }
+    } else {
+      // TODO: Make sure the type is correct.
+      // @ts-ignore
+      target[key] = source[key]
+    }
+  }
+}
+
+export async function resolveMetadata(metadataItems: Item[]) {
+  // TODO: Use `structuredClone` here.
+  let resolvedMetadata: ResolvedMetadata = { ...DEFAULT_METADATA }
+
+  for (const item of metadataItems) {
+    if (item.type === 'layout' || item.type === 'page') {
+      const layerMod = await item.mod()
```

Comment:
If I follow the implementation correctly it seems like each page or layout is a layer (and though you are filtering it out also each file-based metadata like icon.svg)

This doesn't quite follow the semantics of layers as described in the RFC. Each layer is a layer in the filesystem directory structure. Page and Layout share a layer and things like template strings are bound to the layer.

This is important b/c we don't want a title defined in page.js to use a template defined in layout.js in the same directory layer.

I may be misunderstanding the implemenation b/c i'm not super familiar with the technique you are using in next-app-loader (which is super cool btw!) so please forgive me if I'm just not following the code right